res/controllers/Numark-Scratch-scripts.js

```diff
@@ -0,0 +1,452 @@
+            component.disconnect();
+        });
+
+        newMode.forEachComponent(function(component) {
+            component.connect();
+            component.trigger();
+        });
+
+        this.currentMode = newMode;
+    };
+
+    this.currentMode = this.modes[NumarkScratch.PadModeControls.HOTCUE];
+};
+NumarkScratch.PadSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkScratch.ModeHotcue = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.control = NumarkScratch.PadModeControls.HOTCUE;
+
+    this.pads = new components.ComponentContainer();
+    for (var i = 0; i < 4; i++) {
+        this.pads[i] = new components.HotcueButton({
+            group: "[Channel" + deckNumber + "]",
+            midi: [0x93 + deckNumber, 0x14 + i],
+            number: i + 1,
+            shiftControl: true,
+            sendShifted: true,
+            shiftOffset: 0x08,
+            outConnect: false
+        });
+    }
+};
+NumarkScratch.ModeHotcue.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkScratch.ModeRoll = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.control = NumarkScratch.PadModeControls.ROLL;
+
+    this.pads = new components.ComponentContainer();
+    for (var i = 0; i < 4; i++) {
+        this.pads[i] = new components.Button({
+            group: "[Channel" + deckNumber + "]",
+            midi: [0x93 + deckNumber, 0x14 + i],
+            size: NumarkScratch.autoLoopSizes[i],
+            shiftControl: true,
+            sendShifted: true,
+            shiftOffset: 0x08,
+            inKey: "beatlooproll_" + NumarkScratch.autoLoopSizes[i] + "_activate",
+            outKey: "beatloop_" + NumarkScratch.autoLoopSizes[i] + "_enabled",
+            outConnect: false
+        });
+    }
+};
+NumarkScratch.ModeRoll.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkScratch.ModeSampler = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.control = NumarkScratch.PadModeControls.SAMPLER;
+
+    this.pads = new components.ComponentContainer();
+    for (var i = 0; i < 4; i++) {
+        this.pads[i] = new components.SamplerButton({
+            midi: [0x93 + deckNumber, 0x14 + i],
+            number: i + 1,
+            shiftControl: true,
+            sendShifted: true,
+            shiftOffset: 0x08,
+            outConnect: false
+        });
+    }
+};
+NumarkScratch.ModeSampler.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkScratch.Gains = function() {
+    this.cueMix = new components.Pot({
+        group: "[Master]",
+        inKey: "headMix"
+    });
+};
+NumarkScratch.Gains.prototype = new components.ComponentContainer();
+
+NumarkScratch.vuCallback = function(value, group) {
+    var level = value * 90;
+    var deckOffset = script.deckFromGroup(group) - 1;
+    midi.sendShortMsg(0xB0 + deckOffset, 0x1F, level);
+};
```

Comment:
this function will be called very often and `script.deckFromGroup(group)` is very inefficient. Here is a quick tip how you can work around that.

```js
function createVuCallback(deckOffset) {
  return function (value) {
    midi.sendShortMsg(0xB0 + deckOffset, 0x1F, value * 90);
  }
}
// then bind somewhere else like this:
engine.makeConnection("[Channel1]", "VuMeter", createVuCallback(0));
engine.makeConnection("[Channel2]", "VuMeter", createVuCallback(1));
```

This is called a higher-order-function (loosely related to the factory pattern in OOP). This is not an entry-level programming topic, feel free to ask questions if you don't understand my proposal.

Reply:
Done [b5f2633](https://github.com/mixxxdj/mixxx/pull/4834/commits/b5f2633de6e54e7e0da12ced827ecc414b906bc4)
Precommit was complaining about func styles and expecting an expression. Figured it out though. Thanks